project/src/main/puzzle/piece/piece-queue.gd

```diff
@@ -146,18 +146,31 @@ func _fill_remaining_pieces() -> void:
 	while _pieces.size() < MIN_SIZE:
 		# fill a bag with one of each piece and one extra; draw them out in a random order
 		var new_pieces := shuffled_piece_types()
+		_pieces += new_pieces
 		
 		# avoid having two of the same piece consecutively
 		if new_pieces.size() >= 3:
-			for _i in range(48):
-				if _has_duplicate_pieces(new_pieces):
-					new_pieces.shuffle()
-				elif not _pieces.empty() and _pieces.back() == new_pieces[0]:
-					new_pieces.pop_front()
-					new_pieces.insert(int(rand_range(1, new_pieces.size() + 1)), _pieces.back())
-				else:
-					break
-		_pieces += new_pieces
+			# loop over each of the newly added pieces
+			var min_piece_index := _pieces.size() - new_pieces.size()
+			var piece_index := min_piece_index
+			while piece_index < _pieces.size():
+				if _pieces[piece_index] == _pieces[piece_index - 1]:
+					# a piece appears back-to-back in the queue. find a new place for it
```

Comment:
For documentation purposes (and possibly to get away from the pyramid-of-doom here) it might make sense to extract this in a new function; maybe something like `move_duplicate_piece`

Reply:
It would need to return something like `new_position` in the response so that we can futz with `piece_index` appropriately

Reply:
Try it out -- I'm mostly concerned with avoiding having 5-6 levels of indentation as becomes difficult to read and opens the door for bugs (e.g incrementing piece_index conditionally instead of on every iteration of a loop)

Alternatively, we could pull out the entire loop as something like `fix_duplicate_pieces` or something like that.

Reply:
Extracted new function, `func _move_duplicate_piece(from_index: int, min_to_index: int) -> int:`. It wasn't as ugly as I thought.